File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] wires) {
        int answer = n;
        int[][] graph = new int[n+1][n+1];
        
        for (int i = 0; i < wires.length; i++) {
            int a = wires[i][0];
            int b = wires[i][1];
            graph[a][b] = 1;
            graph[b][a] = 1;
        }
        
        for (int i = 0; i < wires.length; i++) {
            int a = wires[i][0];
            int b = wires[i][1];
            graph[a][b] = 0;
            graph[b][a] = 0;
            
            answer = Math.min(answer, bfs(n, a, graph));
            graph[a][b] = 1;
            graph[b][a] = 1;
        }
        
        return answer;
    }
    
    private int bfs(int n, int a, int[][]graph) {
        boolean[] visited = new boolean[n+1];
        visited[a] = true;
        int count = 1;
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(a);
        
        while (!queue.isEmpty()) {
            int node = queue.poll();
            for (int i = 1; i <= n; i++) {
                if (graph[node][i] == 1 && !visited[i]) {
                    visited[i] = true;
                    queue.offer(i);
                    count++;
                }
            }
        }
        return Math.abs(count - (n - count));
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque, defaultdict

def bfs(n, graph, start):
    count = 1
    q = deque([start])
    visited = {i:False for i in range(1, n+1)}
    visited[start] = True

    while q:
        node = q.popleft()
        for next_node in graph[node]:
            if not visited[next_node]:
                visited[next_node] = True
                count += 1
                q.append(next_node)
    
    return abs(count - (n-count))
            
            
def solution(n, wires):
    answer = float('inf')
    
    graph = defaultdict(list)
    
    
    for a, b in wires:
        graph[a].append(b)
        graph[b].append(a)
                
    for a, b in wires:
        # 끊기
        graph[a].remove(b)
        graph[b].remove(a)
    
        answer = min(bfs(n, graph, a), answer)
        
        # 복구
        graph[a].append(b)
        graph[b].append(a)
        
    return answer